### ISSUE TITLE
pkg/http headers should be case-insensitive

### ISSUE THREAD

**Issue by DartBot** [dart-lang/http#13]
_Originally opened as dart-lang/sdk#17062_

_This issue was originally filed by ses...&#064;gmail.com_

---

When using the code snippet below, the content-type header is overloaded by &quot;text/plain&quot;.

&nbsp;&nbsp;import 'package:http/http.dart' as Http;
&nbsp;&nbsp;Http.post(url, body: postData, headers: { &quot;Content-Type&quot;: &quot;text/xml&quot; })

As we can't set _contentType from Request class, could you add 'content-type' in basic client requests also ? (see 'request.dart' source code)

&nbsp;&nbsp;set body(String value) {
&nbsp;&nbsp;&nbsp;&nbsp;bodyBytes = encoding.encode(value);
&nbsp;&nbsp;&nbsp;&nbsp;var contentType = _contentType;
&nbsp;&nbsp;&nbsp;&nbsp;if (contentType == null) {
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;contentType = new ContentType(&quot;text&quot;, &quot;plain&quot;, charset: encoding.name);
&nbsp;&nbsp;&nbsp;&nbsp;} else if (contentType.charset == null) {
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;contentType = new ContentType(contentType.primaryType,
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;contentType.subType,
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;charset: encoding.name,
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;parameters: contentType.parameters);
&nbsp;&nbsp;&nbsp;&nbsp;}
&nbsp;&nbsp;&nbsp;&nbsp;_contentType = contentType;
&nbsp;&nbsp;}

Environement
Dart VM version: 1.1.3 (Thu Feb 06 00:05:52 2014) on &quot;windows_ia32&quot;
http package: http-0.9.2+3

**Comment by DartBot** [comment]
*This comment was originally written by sestegr...&#064;gmail.com*
_____
With the following code snippet, the header 'content-type' is correctly passed.

&nbsp;&nbsp;import 'package:http/http.dart' as Http;
&nbsp;&nbsp;Http.post(url, body: postData, headers: { "content-type": "text/xml" })

Request class shouldn't be case sensitive for HTML headers.

**Comment by DartBot** [comment]
<a href="https://github.com/kevmoo"><img src="https://avatars.githubusercontent.com/u/17034?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [kevmoo](https://github.com/kevmoo)**

----

 cc @nex3.
*Added [Pkg-Http](../labels/Pkg-Http), [Area-Pkg](../labels/Area-Pkg), [Triaged](../labels/Triaged) labels.*

**Comment by DartBot** [comment]
<a href="https://github.com/nex3"><img src="https://avatars.githubusercontent.com/u/188?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [nex3](https://github.com/nex3)**

----

*Changed the title to: "pkg/http headers should be case-insensitive".*

**Comment by DartBot** [comment]
<a href="https://github.com/nex3"><img src="https://avatars.githubusercontent.com/u/188?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [nex3](https://github.com/nex3)**

----

Issue dart-lang/sdk#17405 has been merged into this issue.

**Comment by DartBot** [comment]
<a href="https://github.com/nex3"><img src="https://avatars.githubusercontent.com/u/188?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [nex3](https://github.com/nex3)**

----

r33635 makes the request headers case-insensitive. We'll need to make the response headers case-insensitive before we can release this, though.

**Comment by DartBot** [comment]
<a href="https://github.com/jtmcdole"><img src="https://avatars.githubusercontent.com/u/1924313?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [jtmcdole](https://github.com/jtmcdole)**

----

This breaks multi-part mime boundaries, which are allowed to be in any case they want to be. See:

http://www.w3.org/Protocols/rfc1341/7_2_Multipart.html

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Content-Type: multipart/mixed; 
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;boundary=gc0p4Jq0M2Yt08jU534c0p

which means the boundary markers are:
--gc0p4Jq0M2Yt08jU534c0p
foo bar baz

--gc0p4Jq0M2Yt08jU534c0p

**Comment by DartBot** [comment]
<a href="https://github.com/nex3"><img src="https://avatars.githubusercontent.com/u/188?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [nex3](https://github.com/nex3)**

----

Only the header names are case-insensitive, not the values.

**Comment by donny-dont** [comment]
This is resolved on `master`. Can close this now or with a release.
